Auth0/OAuth2Grant.swift

```diff
@@ -67,50 +81,99 @@ struct PKCE: OAuth2Grant {
-            return callback(.failure(error: WebAuthError.invalidIdTokenNonce))
-        }
-        let clientId = self.authentication.clientId
-        self.authentication
-            .tokenExchange(withCode: code, codeVerifier: verifier, redirectURI: redirectURL.absoluteString)
-            .start { result in
-                // Special case for PKCE when the correct method for token endpoint authentication is not set (it should be None)
-                if case .failure(let cause as AuthenticationError) = result, cause.description == "Unauthorized" {
-                    let error = WebAuthError.pkceNotAllowed("Unable to complete authentication with PKCE. PKCE support can be enabled by setting Application Type to 'Native' and Token Endpoint Authentication Method to 'None' for this app at 'https://manage.auth0.com/#/applications/\(clientId)/settings'.")
-                    callback(Result.failure(error: error))
-                } else {
-                    callback(result)
-                }
+        let idToken = values["id_token"]
+        let responseType = self.responseType
+        let authentication = self.authentication
+        let leeway = self.leeway
+        let nonce = self.defaults["nonce"]
+        let maxAge = self.maxAge
+        let isHybridFlow = responseType.contains(.idToken)
+        let verifier = self.verifier
+        let redirectUrlString = self.redirectURL.absoluteString
+        let clientId = authentication.clientId
+        let validatorContext = IDTokenValidatorContext(authentication: authentication,
+                                                       leeway: leeway,
+                                                       nonce: nonce,
+                                                       maxAge: maxAge)
+        validate(idToken: idToken, for: responseType, with: validatorContext) { error in
+            if let error = error { return callback(.failure(error: error)) }
+            authentication
+                .tokenExchange(withCode: code, codeVerifier: verifier, redirectURI: redirectUrlString)
+                .start { result in
+                    switch result {
+                    case .failure(let error as AuthenticationError) where error.description == "Unauthorized":
+                        // Special case for PKCE when the correct method for token endpoint authentication is not set (it should be None)
+                        let webAuthError = WebAuthError.pkceNotAllowed("Unable to complete authentication with PKCE. PKCE support can be enabled by setting Application Type to 'Native' and Token Endpoint Authentication Method to 'None' for this app at 'https://manage.auth0.com/#/applications/\(clientId)/settings'.")
+                        return callback(.failure(error: webAuthError))
+                    case .failure(let error): return callback(.failure(error: error))
+                    case .success(let credentials):
+                        guard !isHybridFlow else {
```

Comment:
**Nit:** `guard !isHybridFlow else` is confusing ... can you switch this around to be more clear? Like `if this is the hybrid flow then do this thing and return`

Reply:
That's precisely what the `guard` clause is for.

Reply:
Let's negate this conditional and swap the IF's bodies. This way, if front channel is expected, validate and return. "else" (no need to add else) return the received credentials

Reply:
Changed in ca28451be12118c8928745aead45feaa63e669e1.